
    <html>
        <head>
            <title>异常处理</title>
            <meta charset="UTF-8">
            <link rel="stylesheet" href="../index.css">
        </head>
        <body>
        <article class="markdown-body">
    <h1>异常处理</h1>
    <h2 id="异常捕获"><a class="markdown-anchor" href="#异常捕获">#</a> 异常捕获</h2>
<p>得益于框架支持的异步编程模型，错误完全可以用 <code>try catch</code> 来捕获。在编写应用代码时，所有地方都可以直接用 <code>try catch</code> 来捕获异常。</p>
<figure class="highlight js"><table><tr><td class="code"><pre><span class="line"><span class="comment">// app/service/test.js</span></span><br/><span class="line"><span class="keyword">try</span> {</span><br/><span class="line">  <span class="keyword">const</span> res = <span class="keyword">await</span> <span class="keyword">this</span>.ctx.curl(<span class="string">'http://eggjs.com/api/echo'</span>, { <span class="attr">dataType</span>: <span class="string">'json'</span> });</span><br/><span class="line">  <span class="keyword">if</span> (res.status !== <span class="number">200</span>) <span class="keyword">throw</span> <span class="keyword">new</span> <span class="built_in">Error</span>(<span class="string">'response status is not 200'</span>);</span><br/><span class="line">  <span class="keyword">return</span> res.data;</span><br/><span class="line">} <span class="keyword">catch</span> (err) {</span><br/><span class="line">  <span class="keyword">this</span>.logger.error(err);</span><br/><span class="line">  <span class="keyword">return</span> {};</span><br/><span class="line">}</span><br/></pre></td></tr></table></figure>
<p>按照正常代码写法，所有的异常都可以用这个方式进行捕获并处理，但是一定要注意一些特殊的写法可能带来的问题。打一个不太正式的比方，我们的代码全部都在一个异步调用链上，所有的异步操作都通过 await 串接起来了，但是只要有一个地方跳出了异步调用链，异常就捕获不到了。</p>
<figure class="highlight js"><table><tr><td class="code"><pre><span class="line"><span class="comment">// app/controller/home.js</span></span><br/><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">HomeController</span> <span class="keyword">extends</span> <span class="title">Controller</span> </span>{</span><br/><span class="line">  <span class="keyword">async</span> buy () {</span><br/><span class="line">    <span class="keyword">const</span> request = {};</span><br/><span class="line">    <span class="keyword">const</span> config = <span class="keyword">await</span> ctx.service.trade.buy(request);</span><br/><span class="line">    <span class="comment">// 下单后需要进行一次核对，且不阻塞当前请求</span></span><br/><span class="line">    setImmediate(<span class="function"><span class="params">()</span> =&gt;</span> {</span><br/><span class="line">      ctx.service.trade.check(request).catch(<span class="function"><span class="params">err</span> =&gt;</span> ctx.logger.error(err));</span><br/><span class="line">    });</span><br/><span class="line">  }</span><br/><span class="line">}</span><br/></pre></td></tr></table></figure>
<p>在这个场景中，如果 <code>service.trade.check</code> 方法中代码有问题，导致执行时抛出了异常，尽管框架会在最外层通过 <code>try catch</code> 统一捕获错误，但是由于 <code>setImmediate</code> 中的代码『跳出』了异步链，它里面的错误就无法被捕捉到了。因此在编写类似代码的时候一定要注意。</p>
<p>当然，框架也考虑到了这类场景，提供了 <code>ctx.runInBackground(scope)</code> 辅助方法，通过它又包装了一个异步链，所有在这个 scope 里面的错误都会统一捕获。</p>
<figure class="highlight js"><table><tr><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">HomeController</span> <span class="keyword">extends</span> <span class="title">Controller</span> </span>{</span><br/><span class="line">  <span class="keyword">async</span> buy () {</span><br/><span class="line">    <span class="keyword">const</span> request = {};</span><br/><span class="line">    <span class="keyword">const</span> config = <span class="keyword">await</span> ctx.service.trade.buy(request);</span><br/><span class="line">    <span class="comment">// 下单后需要进行一次核对，且不阻塞当前请求</span></span><br/><span class="line">    ctx.runInBackground(<span class="keyword">async</span> () =&gt; {</span><br/><span class="line">      <span class="comment">// 这里面的异常都会统统被 Backgroud 捕获掉，并打印错误日志</span></span><br/><span class="line">      <span class="keyword">await</span> ctx.service.trade.check(request);</span><br/><span class="line">    });</span><br/><span class="line">  }</span><br/><span class="line">}</span><br/></pre></td></tr></table></figure>
<p><strong>为了保证异常可追踪，必须保证所有抛出的异常都是 Error 类型，因为只有 Error 类型才会带上堆栈信息，定位到问题。</strong></p>
<h2 id="框架层统一异常处理"><a class="markdown-anchor" href="#框架层统一异常处理">#</a> 框架层统一异常处理</h2>
<p>框架通过 <a href="https://github.com/eggjs/egg-onerror" target="_blank" rel="noopener">onerror</a> 插件提供了统一的错误处理机制。对一个请求的所有处理方法（Middleware、Controller、Service）中抛出的任何异常都会被它捕获，并自动根据请求想要获取的类型返回不同类型的错误（基于 <a href="https://tools.ietf.org/html/rfc7231#section-5.3.2" target="_blank" rel="noopener">Content Negotiation</a>）。</p>
<table>
<thead>
<tr>
<th>请求需求的格式</th>
<th>环境</th>
<th>errorPageUrl 是否配置</th>
<th>返回内容</th>
</tr>
</thead>
<tbody>
<tr>
<td>HTML &amp; TEXT</td>
<td>local &amp; unittest</td>
<td>-</td>
<td>onerror 自带的错误页面，展示详细的错误信息</td>
</tr>
<tr>
<td>HTML &amp; TEXT</td>
<td>其他</td>
<td>是</td>
<td>重定向到 errorPageUrl</td>
</tr>
<tr>
<td>HTML &amp; TEXT</td>
<td>其他</td>
<td>否</td>
<td>onerror 自带的没有错误信息的简单错误页（不推荐）</td>
</tr>
<tr>
<td>JSON &amp; JSONP</td>
<td>local &amp; unittest</td>
<td>-</td>
<td>JSON 对象或对应的 JSONP 格式响应，带详细的错误信息</td>
</tr>
<tr>
<td>JSON &amp; JSONP</td>
<td>其他</td>
<td>-</td>
<td>JSON 对象或对应的 JSONP 格式响应，不带详细的错误信息</td>
</tr>
</tbody>
</table>
<h3 id="errorpageurl"><a class="markdown-anchor" href="#errorpageurl">#</a> errorPageUrl</h3>
<p>onerror 插件的配置中支持 errorPageUrl 属性，当配置了 errorPageUrl 时，一旦用户请求线上应用的 HTML 页面异常，就会重定向到这个地址。</p>
<p>在 <code>config/config.default.js</code> 中</p>
<figure class="highlight js"><table><tr><td class="code"><pre><span class="line"><span class="comment">// config/config.default.js</span></span><br/><span class="line"><span class="built_in">module</span>.exports = {</span><br/><span class="line">  onerror: {</span><br/><span class="line">    <span class="comment">// 线上页面发生异常时，重定向到这个页面上</span></span><br/><span class="line">    errorPageUrl: <span class="string">'/50x.html'</span>,</span><br/><span class="line">  },</span><br/><span class="line">};</span><br/></pre></td></tr></table></figure>
<h2 id="自定义统一异常处理"><a class="markdown-anchor" href="#自定义统一异常处理">#</a> 自定义统一异常处理</h2>
<p>尽管框架提供了默认的统一异常处理机制，但是应用开发中经常需要对异常时的响应做自定义，特别是在做一些接口开发的时候。框架自带的 onerror 插件支持自定义配置错误处理方法，可以覆盖默认的错误处理方法。</p>
<figure class="highlight js"><table><tr><td class="code"><pre><span class="line"><span class="comment">// config/config.default.js</span></span><br/><span class="line"><span class="built_in">module</span>.exports = {</span><br/><span class="line">  onerror: {</span><br/><span class="line">    all(err, ctx) {</span><br/><span class="line">      <span class="comment">// 在此处定义针对所有响应类型的错误处理方法</span></span><br/><span class="line">      <span class="comment">// 注意，定义了 config.all 之后，其他错误处理方法不会再生效</span></span><br/><span class="line">      ctx.body = <span class="string">'error'</span>;</span><br/><span class="line">      ctx.status = <span class="number">500</span>;</span><br/><span class="line">    },</span><br/><span class="line">    html(err, ctx) {</span><br/><span class="line">      <span class="comment">// html hander</span></span><br/><span class="line">      ctx.body = <span class="string">'&lt;h3&gt;error&lt;/h3&gt;'</span>;</span><br/><span class="line">      ctx.status = <span class="number">500</span>;</span><br/><span class="line">    },</span><br/><span class="line">    json(err, ctx) {</span><br/><span class="line">      <span class="comment">// json hander</span></span><br/><span class="line">      ctx.body = { <span class="attr">message</span>: <span class="string">'error'</span> };</span><br/><span class="line">      ctx.status = <span class="number">500</span>;</span><br/><span class="line">    },</span><br/><span class="line">    jsonp(err, ctx) {</span><br/><span class="line">      <span class="comment">// 一般来说，不需要特殊针对 jsonp 进行错误定义，jsonp 的错误处理会自动调用 json 错误处理，并包装成 jsonp 的响应格式</span></span><br/><span class="line">    },</span><br/><span class="line">  },</span><br/><span class="line">};</span><br/></pre></td></tr></table></figure>
<h2 id="404"><a class="markdown-anchor" href="#404">#</a> 404</h2>
<p>框架并不会将服务端返回的 404 状态当做异常来处理，但是框架提供了当响应为 404 且没有返回 body 时的默认响应。</p>
<ul>
<li>
<p>当请求被框架判定为需要 JSON 格式的响应时，会返回一段 JSON：</p>
<figure class="highlight json"><table><tr><td class="code"><pre><span class="line">{ <span class="attr">"message"</span>: <span class="string">"Not Found"</span> }</span><br/></pre></td></tr></table></figure>
</li>
<li>
<p>当请求被框架判定为需要 HTML 格式的响应时，会返回一段 HTML：</p>
<figure class="highlight html"><table><tr><td class="code"><pre><span class="line"><span class="tag">&lt;<span class="name">h1</span>&gt;</span>404 Not Found<span class="tag">&lt;/<span class="name">h1</span>&gt;</span></span><br/></pre></td></tr></table></figure>
</li>
</ul>
<p>框架支持通过配置，将默认的 HTML 请求的 404 响应重定向到指定的页面。</p>
<figure class="highlight js"><table><tr><td class="code"><pre><span class="line"><span class="comment">// config/config.default.js</span></span><br/><span class="line"><span class="built_in">module</span>.exports = {</span><br/><span class="line">  notfound: {</span><br/><span class="line">    pageUrl: <span class="string">'/404.html'</span>,</span><br/><span class="line">  },</span><br/><span class="line">};</span><br/></pre></td></tr></table></figure>
<h3 id="自定义-404-响应"><a class="markdown-anchor" href="#自定义-404-响应">#</a> 自定义 404 响应</h3>
<p>在一些场景下，我们需要自定义服务器 404 时的响应，和自定义异常处理一样，我们也只需要加入一个中间件即可对 404 做统一处理：</p>
<figure class="highlight js"><table><tr><td class="code"><pre><span class="line"><span class="comment">// app/middleware/notfound_handler.js</span></span><br/><span class="line"><span class="built_in">module</span>.exports = <span class="function"><span class="params">()</span> =&gt;</span> {</span><br/><span class="line">  <span class="keyword">return</span> <span class="keyword">async</span> <span class="function"><span class="keyword">function</span> <span class="title">notFoundHandler</span>(<span class="params">ctx, next</span>) </span>{</span><br/><span class="line">    <span class="keyword">await</span> next();</span><br/><span class="line">    <span class="keyword">if</span> (ctx.status === <span class="number">404</span> &amp;&amp; !ctx.body) {</span><br/><span class="line">      <span class="keyword">if</span> (ctx.acceptJSON) {</span><br/><span class="line">        ctx.body = { <span class="attr">error</span>: <span class="string">'Not Found'</span> };</span><br/><span class="line">      } <span class="keyword">else</span> {</span><br/><span class="line">        ctx.body = <span class="string">'&lt;h1&gt;Page Not Found&lt;/h1&gt;'</span>;</span><br/><span class="line">      }</span><br/><span class="line">    }</span><br/><span class="line">  };</span><br/><span class="line">};</span><br/></pre></td></tr></table></figure>
<p>在配置中引入中间件：</p>
<figure class="highlight js"><table><tr><td class="code"><pre><span class="line"><span class="comment">// config/config.default.js</span></span><br/><span class="line"><span class="built_in">module</span>.exports = {</span><br/><span class="line">  middleware: [ <span class="string">'notfoundHandler'</span> ],</span><br/><span class="line">};</span><br/></pre></td></tr></table></figure>

  </article>
        </body>
    </html>